src/hotspot/share/opto/phaseX.cpp

```diff
@@ -469,6 +469,13 @@ PhaseRenumberLive::PhaseRenumberLive(PhaseGVN* gvn,
 
   uint worklist_size = worklist->size();
 
+  GrowableArray<Node_Notes*>* old_node_note_array = C->node_note_array();
+  if (old_node_note_array != nullptr) {
+    int new_size = (_useful.size() >> 8) + 1; // The node note array uses blocks, see C->_log2_node_notes_block_size
```

Comment:
You should call `new_size = MAX2(8, new_size)` to make sure that we have at least 8 elements for initial allocation.

Reply:
Okay, I added that. The 8 seems arbitrary to me but since we already use that for initial allocation of the array, we can as well be consistent here. Just note that since we are calling `C->grow_node_notes`, we will also initialize with `Node_Notes*` right away.

Reply:
Why don't we just use `C->_log2_node_notes_block_size` directly in (_useful.size() >> 8)? 

I don't understand why we have to add MAX2(8, new_size) either. It looks like c2 doesn't want to have node-level accuracy. It drops the lowest 8bits of node_idx as block_id. I think the minimal number of "block" is 1, or arr is NULL.

Reply:
Thanks for the clarification. yes, I understanding was wrong. 

A block is an array of 256 Node_Note.  Is it a particular reason that Compile needs at least 8 blocks?  It can grow automatically anyway.

Reply:
> Is it a particular reason that Compile needs at least 8 blocks?

It's the value that is used when creating the array:
https://github.com/openjdk/jdk/blob/4619e8bae838abd1f243c2c65a538806d226b8e8/src/hotspot/share/opto/compile.cpp#L1060-L1062

I don't think there is a particular reason for 8 but it's just one of those more or less reasonable default values/sizes that we use all over the place when creating containers.